test/hotspot/jtreg/compiler/c2/irTests/ProfileAtTypeCheck.java

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (c) 2023, Red Hat, Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Platform;
+import java.util.ArrayList;
+
+/*
+ * @test
+ * bug 8308869
+ * @summary C2: use profile data in subtype checks when profile has more than one class
+ * @library /test/lib /
+ * @requires vm.compiler2.enabled
+ * @run driver compiler.c2.irTests.ProfileAtTypeCheck
+ */
+
+public class ProfileAtTypeCheck {
+    public static void main(String[] args) {
+        ArrayList<String> flags = new ArrayList<>();
+        flags.add("-XX:TypeProfileSubTypeCheckCommonThreshold=90");
+        if (!Platform.is32bit()) {
+            flags.add("-XX:-UseCompressedClassPointers");
+        }
+        flags.add("-XX:+IgnoreUnrecognizedVMOptions");
+        flags.add("-XX:+UseParallelGC");
```

Comment:
The test fails when another GC is specified externally.

Reply:
Should be fixed in the new commit. I also improved the test a bit (added one test case with array store checks, duplicated the rules so there's no need to run with -XX:-UseCompressedClassPointers, run tests with only c1 profile collection and only interpreter profile collection to exercise new profile collection code paths independently).